### PR TITLE
Rebrand "Mastercoin" to "Omni"

### PR DESCRIPTION
**Pull request by zathras-crypto**
The foundation has requested that the SP1 and SP2 tokens are rebranded from Mastercoin to Omni (MSC => OMNI).

Only user facing adjustments have been made to keep the scope limited.  Comments, functions and variable names have not been modified.

### CODE REVIEW

File: src/qt/lookupspdialog.cpp [hunk 1]
```diff
@@ -283,12 +283,12 @@ void LookupSPDialog::updateDisplayedProperty()
     }
     // overrides for MSC and TMSC - temporary code can be removed when metadata is modified
```
Comment (dexX7) [inline src/qt/lookupspdialog.cpp]: This can probably be removed now. The hardcoded SPs should get a description.

**Comment by zathras-crypto** [comment]
Notes
* The protocol does not support renaming of properties at this stage.  
* The argument has been made that since MSC & TMSC were not created by the protocol and are instead hardcoded that this limitation does not apply.
* The final decision for this PR is probably for the board?

File: src/qt/lookupspdialog.cpp (hunk 1 above)
Comment (zathras-crypto) [inline src/qt/lookupspdialog.cpp]: Good point, thanks :)

**Comment by dexX7** [comment]
It would be great, if you could quickly update the API docs as well.

> The foundation has requested ... The final decision for this PR is probably for the board?

This is a bit contradicting (at least for me), because It's not clear to me, whether this was approved or not, and if so, why would it need another decision then?

> TOMNI ...

This almost looks like the name TOMMY. >_<"

**Comment by zathras-crypto** [comment]
> This is a bit contradicting (at least for me), because It's not clear to me, whether this was approved or not, and if so, why would it need another decision then?

I'm not 100% on *who* get's to make that decision.  For some background, @patrickdugan was talking about an (already underway) rebranding effort with exchanges to rename the token.  I raised that our reference (Omni Core) didn't have this rebranding of the MSC token & also that I hadn't been asked to make these changes in OmniChest.info.  

I thus said I would submit a PR and asked everyone to please come and make their comments re rebranding the token on the PR here.

As I understood it from the meeting the rebrand of the token is already in full swing with third parties, so that would indicate at least tacit approval, but to be honest I'm a little shaky on authority/who is actually on the 'board' these days :(

>> TOMNI

>  This almost looks like the name TOMMY. >_<

Hehe yeah it'll take some getting used to!

**Comment by CraigSellars** [comment]
We approved the name change, glad to see it’s moving forward.  Thanks, guys!

**Comment by dexX7** [comment]
The tests currently fail because of the following OmniJ test:

- https://github.com/OmniLayer/OmniJ/blob/master/omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/basic/ListPropertiesSpec.groovy#L28

This was forseeable. I'm going to visually check the submission once more later, but so far I have no objections.

**Comment by patrickdugan** [comment]
Hey guys, I take it we just need the repo admin to approve this and the follow-up PR and we're good. 

Quick question, will these changes show up in the 0.9.99 test build we are promoting as well? If not, could we get a PR for that branch as well so the experience is consistent in our tester population?

**Comment by zathras-crypto** [comment]
> Hey guys, I take it we just need the repo admin to approve this and the follow-up PR and we're good. 

@CraigSellars has approved the change (see couple of comments up)

> Quick question, will these changes show up in the 0.9.99 test build we are promoting as well?

These changes will show up in 0.0.9.99 test builds once:
* This PR is merged
* New 0.0.9.99 test builds are created (both @dexX7 & I have created new builds recently but trying to get to the bottom of a hash mismatch for the windows setup executables)
* New test builds are uploaded to web URLs users are downloading from

**Comment by dexX7** [comment]
If possible, I'd like to get the feature activation and send-all transaction type into the next preview, because both seem to be ready, or almost ready, and more exposure would be great.

This isn't required though, and I'd be fine with another preview before that, too.

Unfortunally I don't see many options to resolve the issue with the mismatching installers, so I tend to just exclude the installers, and simply publish the other files (which match perfectly).

**Comment by dexX7** [comment]
The new hardcoded descriptions currently break the tests, which is really aweful. I'm going to push a quick fix for OmniJ, and once that was adopted, I'm fine with merging this PR.

**Comment by patrickdugan** [comment]
When you're done with the tests, push just this change so I can send screen caps to the exchange partners and prompt them to push their changes. I will start promoting this test build then on those exchanges to make the jump. We should try to get the two features you mention in for next week's preview.

**Comment by dexX7** [comment]
@zathras-crypto: quick note regarding the old previous discussion: I think we should add `http://`, if that's okay with you. And maybe add a `.` at the end of the property description.

**Comment by dexX7** [comment]
@patrickdugan: can you please clarify what you mean with:

> and prompt them to push their changes

?

I do not recommend to run the test builds in production environments, and if there is any risk an exchange does it nevertheless, we are basically forced to include the outstanding consensus affecting changes in the next preview.

Somewhat related information: we are usually able to build and compare preview builds within a day.

**Comment by zathras-crypto** [comment]
> I will start promoting this test build then on those exchanges to make the jump.

Just for 100% clarity, this change is for 0.0.10 (currently being previewed as 0.0.9.99) - we would *not* want any exchanges to "make the jump" to one of our 0.0.10 preview builds, they are not release-ready yet.

> If possible, I'd like to get the feature activation and send-all transaction type into the next preview, because both seem to be ready, or almost ready, and more exposure would be great.

Agreed, they are 99% ready so it seems prudent to include them.

> Unfortunally I don't see many options to resolve the issue with the mismatching installers, so I tend to just exclude the installers, and simply publish the other files (which match perfectly).

The installers are beneficial, and since these are preview builds perhaps it's OK that we have a non-deterministic element?  Also to remember is (for whatever reason) I can create build after build with the same setup hash on my Gitian VM.

> ...push just this change so I can send screen caps to the exchange partners and prompt them to push their changes....

Should we not be organizing this a little more?  Eg a PR effort; corresponding blog posts, reddit posts etc etc - given OmniChest.info is the "goto" block explorer for Omni that is where most people are going to get visibility into the MSC -> OMNI rebrand, so as mentioned yesterday at the meeting changes should be made there too.  

TL:DR; rather than just "pulling the pin" as it were, perhaps better to structure the change and include as many stakeholders as we can?  IMO the 'change' should go 'live' at the same time for:

* PR effort (blog, reddit, skype groups etc etc)
* Block explorer (OmniChest)
* Web wallet (OmniWallet)
* Exchanges (as many as we can agree scheduling with)

Just my 2 cents :)

**Comment by zathras-crypto** [comment]
> @zathras-crypto: quick note regarding the old previous discussion: I think we should add http://, if that's okay with you. And maybe add a . at the end of the property description.

Sure dude, will do :)

**Comment by zathras-crypto** [comment]
> screen caps

As requested, but it's critical to note these are screen caps from a development version of Omni Core, and intergrators should not go and run this code :)

screenshot from 2015-09-03 12 22 00
screenshot from 2015-09-03 12 22 22
screenshot from 2015-09-03 12 21 55
screenshot from 2015-09-03 12 27 19

**Comment by msgilligan** [comment]
FYI @zathras-crypto and @patrickdugan (as I commented on Dexx's OmniJ PR https://github.com/OmniLayer/OmniJ/issues/99)

Although the Goal of OmniJ release 0.3 is to provide an RPC client and transaction builder that works with 0.0.9 **and** 0.0.10, the integration tests (RegTest and Consensus) are focused on 0.0.10 only (mostly)  These priorities are simply based upon what people are using OmniJ for. The only thing that I know of that is being tested is 0.0.10. Whereas the other OmniJ functions are being used by multiple projects that are currently working with 0.0.9.

I'm open for feedback on these priorities -- especially if there are use cases that I'm not aware of.

**Comment by dexX7** [comment]
@zathras-crypto: I share your sentiment, and I think it would be helpful to lay out a plan.

I'm more than split regarding:

> The installers are beneficial, and since these are preview builds perhaps it's OK that we have a non-deterministic element? Also to remember is (for whatever reason) I can create build after build with the same setup hash on my Gitian VM.

While we are not yet at a point, where it's very user friendly to verify deterministic builds, I believe breaking the approach, of having at least two matching results, undermines the whole process almost completely. So I tend to believe it would be better to leave out the non-matching elements (and of course, ideally resolving the initial issue of the mismatches).

@msgilligan: thanks for your clarification!

**Comment by zathras-crypto** [comment]
> I believe breaking the approach, of having at least two matching results, undermines the whole process almost completely

That's a very fair point, and I don't think we'll lose many potential testers due to the lack of a full windows installer (it's not like we have to install a bunch of dependencies, just extracting the QT binary is all that's needed).  OK, so we'll do the next preview without the windows installers in the case that we're not able to resolve whatever the issue is prior to building.

@msgilligan thanks, you may be interested in looking at consensus testing via consensus hash in OmniJ when you have spare time as it should simplify the process substantially.  Support via RPC (```omni_getcurrentconsensushash```) is already in the current 0.0.10 codebase and the next version of the OmniChest API has support for it.  Not sure how that fits into your priorities but let me know if I can help at all :)

While on the topic @dexX7 what do you think about expanding the scope of the consensus hash to include additional state attributes?  If so how far should we go?  For example to confirm SPs do we include all properties with their metadata, or is it sufficient to just include the ```next_spid``` and ```next_test_spid``` vars?  I'd also like to include data from all other values that affect the current state such as open positions (DEx offers/accepts, open MetaDEx trades, crowdsales etc) since that is all data that can be used to actually verify consensus of the state (rather than verifying consensus of the balances which only infers consensus of the state).

**Comment by dexX7** [comment]
@zathras-crypto: yes, we should expand the scope of the consensus hash.

A few notes here, first, let's make all consensus hashes unique:

- currently we generate each consensus hash with the tally for a specific block
- however, if two or more tally-collections are similar, then so are their hashes
- this means there might very well be more than one similar consensus hash
- what if we combine the previous consensus hash with the tally hash?
- say: `consensus_hash[block n] = Hash(consensus_hash[block n-1] + tally_hash[block n])`
- for the very first hash, there won't be a predecessor, so we'd need a filler
- this is probably tricky during reorgs!

Expand the tally hash to operations, and not results:

- currently we look at the results at the end of the block
- but we basically don't care what happens inbetween
- there might be similar balances, but with a different history
- e.g. `10 MSC + 15 MSC == 7 MSC + 18 MSC`
- what if we expand the consensus hash on each balance changing operation?
- this is much more tricky, because we batch-restore balances via persistence as sum, and not as single operations

Then you already mentioned:

- we could also cover other values
- such as: transactions, smart properties, open orders, ...
- while transactions are somewhat easy to represent, we'd need a representation for these other items

I would like all this very much, but the more complex the consensus hashing becomes, the more pitfalls there probably are. It would also be great, if the consensus hash generation isn't intrusive, e.g. I wouldn't like to hijack to very low tally operations like `CMPTally::updateMoney`.

**Comment by zathras-crypto** [comment]
> * this means there might very well be more than one similar consensus hash
> * what if we combine the previous consensus hash with the tally hash?
> * for the very first hash, there won't be a predecessor, so we'd need a filler
> * this is probably tricky during reorgs!

I'm not sure about this one - it boils down to whether consensus hashes should be atomic or chained.  If atomic then the state can be verified without knowledge of past history, if chained then we must know consensus hash for the previous block in order to verify current, and in order to verify the previous consensus hash we need the hash for the block before that, and so on working backwards.

I can see pros and cons to both approaches - do you have a view here?

Note, if the objective is purely to make the hashes unique, adding the blockhash to the hash would serve the same purpose without requiring prior knowledge if we feel atomic consensus hashes are desired.

> Expand the tally hash to operations, and not results:

Just a thought, but perhaps there is a half-way option - if per block isn't granular enough, why not expand the hash to per transaction.  If the hash was expanded to cover much more of that state then any transaction that changed any aspect of the state in a transaction would result in a different hash after that transaction.  I raise it as that would require very minimal changes I think? 

> I would like all this very much, but the more complex the consensus hashing becomes, the more pitfalls there probably are

Yeah, I was thinking to keep the changes within ```GetConsensusHash()``` which we can call when we decide to (on operation, on transaction, on block etc)...   But I do see your point, and we shouldn't go crazy hehe :)

**Comment by msgilligan** [comment]
@zathras-crypto I'm definitely interested in using checking consensus hashes in the tests. I should also be able to write a test that verifies the consensus hashes, right? Is the algorithm documented somewhere?

**Comment by zathras-crypto** [comment]
@sean FYI (& you can thank @dexX7 for the better commenting, my commenting wasn't that clear when I originally wrote the function hehe):

```
/**
 * Obtains a hash of all balances to use for consensus verification and checkpointing.
 *
 * For increased flexibility, so other implementations like OmniWallet and OmniChest can
 * also apply this methodology without necessarily using the same exact data types (which
 * would be needed to hash the data bytes directly), create a string in the following
 * format for each address/property identifier combo to use for hashing:
 *
 * Format specifiers:
 *   "%s|%d|%d|%d|%d|%d"
 *
 * With placeholders:
 *   "address|propertyid|balance|selloffer_reserve|accept_reserve|metadex_reserve"
 *
 * Example:
 *   SHA256 round 1: "1LCShN3ntEbeRrj8XBFWdScGqw5NgDXL5R|31|8000|0|0|0"
 *   SHA256 round 2: "1LCShN3ntEbeRrj8XBFWdScGqw5NgDXL5R|2147483657|0|0|0|234235245"
 *   SHA256 round 3: "1LP7G6uiHPaG8jm64aconFF8CLBAnRGYcb|1|0|100|0|0"
 *   SHA256 round 4: "1LP7G6uiHPaG8jm64aconFF8CLBAnRGYcb|2|0|0|0|50000"
 *   SHA256 round 5: "1LP7G6uiHPaG8jm64aconFF8CLBAnRGYcb|10|0|0|1|0"
 *   SHA256 round 6: "3CwZ7FiQ4MqBenRdCkjjc41M5bnoKQGC2b|1|12345|5|777|9000"
 *
 * Result:
 *   "3580e94167f75620dfa8c267862aa47af46164ed8edaec3a800d732050ec0607"
 *
 * The byte order is important, and in the example we assume:
 *   SHA256("abc") = "ad1500f261ff10b49c7a1796a36103b02322ae5dde404141eacf018fbf1678ba"
 *
 * Note: empty balance records and the pending tally are ignored. Addresses are sorted based
 * on lexicographical order, and balance records are sorted by the property identifiers.
 */
uint256 GetConsensusHash() {
```

**Comment by dexX7** [comment]
@msgilligan: @zathras-crypto copied the relevant part from here, and the unit test may be useful in this context:

- [src/omnicore/rules.cpp#L435-L507](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/rules.cpp#L435-L507)
- [src/omnicore/test/checkpoint_tests.cpp](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/test/checkpoint_tests.cpp)
- [src/omnicore/rpc.cpp#L1561-L1593](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/rpc.cpp#L1561-L1593)

> I can see pros and cons to both approaches - do you have a view here?

You have a very good point. I've never considered atomic consensus hashes as desirable, but actually, when thinking about it, it has indeed benefits, especially because we even restore state snapshots. On the other hand, at least now, we use the consensus hashes to checkcorrect behavior, to pin down issues, and having the whole history "baked in" would cover "everything".

**Comment by zathras-crypto** [comment]
@msgilligan @dexX7 I made some changes to consensus hashing here https://github.com/OmniLayer/omnicore/pull/226 based on our discussions.  I still haven't put in a "predecessor" hash yet, still unsure which way to go here.
